### DISCUSSION TITLE
Use swift commands in SPM projects

### DISCUSSION THREAD

**Discussion by m-housh** [wojciech-kulik/xcodebuild.nvim#248] [Ideas]
It would be nice if building and testing in an SPM package used `swift build` and `swift test` instead of the `xcodebuild` commands that require a scheme and workspace.

The downside of using `xcodebuild` (which very well may be my skill issue) is having to constantly open `Xcode` and update the package scheme when adding new targets whereas `swift build` and `swift test` generally just work as expected.

When working inside of `Xcode` I believe this is managed for you / new targets get added to the default `<project>-Package` scheme, however when I'm working in `nvim` this doesn't occur, which leads to a couple of issues. Firstly LSP doesn't really work until you add the target to the scheme and build and secondly your project still builds however you may not be building all the targets that you think you are.

At any rate, this is a great project and I really appreciate your work.

**Comment by wojciech-kulik** [comment]
Hi,

Interesting case, the problem with using `swift` instead of `xcodebuild` is that it's missing things that the plugin requires to add some extra functionality.

For example, Test Explorer requires knowledge about tests received from logs, derived data, and `xcresult` file. Based on that all marks, diagnostics, code coverage, navigation to tests, and test tree are prepared. However, `swift test` doesn't provide most of those things. Not much more would have remained in general, in that case you could just bind `swift test` without all those integrations. 

Regarding the build command, I would need to investigate that, because I'm not sure yet.

> When working inside of Xcode I believe this is managed for you / new targets get added to the default <project>-Package scheme, however when I'm working in nvim this doesn't occur

Hmm, I will check that, maybe it requires some flag or extra command.

**Reply by m-housh** [reply]
I have keybindings setup to run both `swift build` and `swift test` in a pop-over terminal, using a plugin I started on a while ago but never got much time fully flesh it out, but at the time I started it the LSP integration was barely usable, so I went back to Xcode although I would much prefer to work inside of neovim and my terminal.

I've only been experimenting with your plugin for a few days, so this may just be a me thing. But I would also expect that a solution for SPM projects and that doesn't rely on `Xcode` would make this plugin more useful to folks who would be developing on Linux or Windows machines (i.e. swift-server apps).

Since adding your plugin to my config and playing with it for a few days, it's made me think that development using something other than Xcode is much closer than I thought as opposed to the last time I tried.

Keep up the great work, I appreciate it!

**Reply by wojciech-kulik** [reply]
The plugin was initially meant for Xcode project development (it's called xcodebuild.nvim - as it was supposed to be a wrapper for xcodebuild commands), I just added SPM support about 2 weeks ago maybe :D. 

However, if it is possible to replace `xcodebuild` commands with `swift` commands for SPM development and reduce system-specific dependencies, then sure, it would be definitely a good idea. Although, I need to investigate what limitations and issues there might be.

Do you have any problems with running tests using `xcodebuild`? As I understand the main issue you have is with building because new targets are not included.

**Reply by m-housh** [reply]
Turns out that the building and testing issues I was having was because I was using the system wide LSP `/usr/bin/sourcekit-lsp`.  When I switched my lsp config to use [per this discussion](https://github.com/wojciech-kulik/xcodebuild.nvim/discussions/193#discussioncomment-10849293), it works as intended.

**Reply by wojciech-kulik** [reply]
So do I understand correctly that right now everything works as expected on your side? :)

**Reply by m-housh** [reply]
Yes, but I do sometimes get inconsistent showings in the test explorer, showing an [x] although the test passes. This test target uses `Swift-Testing` as opposed to `XCTest` for my integration tests.

![Screenshot 2024-11-18 at 8 23 18 AM](https://github.com/user-attachments/assets/d5e3b9f3-cab3-48ea-9817-382219f63935)
![Screenshot 2024-11-18 at 8 23 30 AM](https://github.com/user-attachments/assets/87a9a685-157e-4e43-834c-1e880e3786cf)

**Reply by m-housh** [reply]
And actually to clarify upon further investigation the test did fail with one of the arguments, so test explorer was showing correctly, however the logs did not show a failing a test.

**Reply by wojciech-kulik** [reply]
Could you share `.nvim/xcodebuild/original_logs.log`?

**Reply by m-housh** [reply]
Here's the logs reproducing the failing test.

[original_logs.log](https://github.com/user-attachments/files/17803842/original_logs.log)
[xcodebuild.log](https://github.com/user-attachments/files/17803845/xcodebuild.log)

**Reply by wojciech-kulik** [reply]
@m-housh could you switch to that branch: #250 and see if it works as expected now? I used your logs in snapshot tests, I hope you don't mind. I obfuscated any ids/names. If you notice anything else that should be obfuscated, let me know.

**Reply by m-housh** [reply]
They were all bogus names / id's for testing purposes. I had to switch to using XCTest because Swift-Testing doesn't work well with the packages I need inside of docker / Linux, but I'll setup another test package when I have time and see if that branch solves the problem and report back.

**Reply by m-housh** [reply]
I still get the same result when updating to that branch, the explorer shows correctly but the summary and notifications say all tests passed.  Here's the log files.

<img width="701" alt="Screenshot 2024-11-18 at 5 46 41 PM" src="https://github.com/user-attachments/assets/af09b911-8e0a-4267-b88d-0debc331c335">

**Reply by m-housh** [reply]
Logs that seemed to not upload.

[original_logs.log](https://github.com/user-attachments/files/17807110/original_logs.log)
[xcodebuild.log](https://github.com/user-attachments/files/17807111/xcodebuild.log)

**Reply by wojciech-kulik** [reply]
Hmm, did you restart Neovim after updating the plugin? Because your logs are parsed correctly and the final message is printed based on those logs, so I'm not sure what could cause that issue.

**Reply by wojciech-kulik** [reply]
I've got one more idea, could you attach `report.json`?

**Reply by m-housh** [reply]
Yes, I restarted neovim. I also cleaned my DerivedData folder to see if it that had something to do with it. Deleted the .nvim/xcodebuild files and setup again and same results. Here's the report.json.

[report.json](https://github.com/user-attachments/files/17807433/report.json)

**Reply by wojciech-kulik** [reply]
Oh, I forgot that at the end results are adjusted based on xcresult file, maybe there is a bug.

**Reply by wojciech-kulik** [reply]
I pushed one more fix, please checkout latest changes and see if it helps

**Reply by m-housh** [reply]
[xcresulttool.out.json](https://github.com/user-attachments/files/17807584/xcresulttool.out.json)

**Reply by m-housh** [reply]
It's working in this commit.

<img width="701" alt="Screenshot 2024-11-18 at 7 22 19 PM" src="https://github.com/user-attachments/assets/bdbf6699-4efb-41d4-a35f-348ea5c1910e">

**Reply by wojciech-kulik** [reply]
Great :) If you encounter any other issues feel free to create a new issue.

**Reply by wojciech-kulik** [reply]
Can I close this discussion or is there anything else?

**Reply by m-housh** [reply]
Nothing else here, thanks!
